lib/nrf_modem_lib/nrf91_net_l2/nrf91_net_l2.c

```diff
@@ -0,0 +1,328 @@
+
+	int ret = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_DEACTIVATE_LTE);
+
+	if (ret) {
+		LOG_ERR("lte_lc_func_mode_set, error: %d", ret);
+		net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+		return;
+	}
+
+	net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_TIMEOUT, iface_bound);
+}
+
+static int modem_init(void)
+{
+	LOG_DBG("Initializing nRF Modem Library");
+
+	int ret = nrf_modem_lib_init(NORMAL_MODE);
+
+	/* Handle return values related to modem DFU. */
+	switch (ret) {
+	case 0:
+		/* Initialization successful, no action required. */
+		return 0;
+	case NRF_MODEM_DFU_RESULT_OK:
+		LOG_DBG("Modem DFU successful. The modem will run the updated "
+			"firmware after reinitialization.");
+		break;
+	case NRF_MODEM_DFU_RESULT_UUID_ERROR:
+	case NRF_MODEM_DFU_RESULT_AUTH_ERROR:
+		LOG_ERR("Modem DFU error: %d. The modem will automatically run the previous "
+			"(non-updated) firmware after reinitialization.", ret);
+		break;
+	case NRF_MODEM_DFU_RESULT_VOLTAGE_LOW:
+		LOG_ERR("Modem DFU not executed due to low voltage, error: %d. "
+			"The modem will retry the update on reinitialization.", ret);
+		break;
+	case NRF_MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case NRF_MODEM_DFU_RESULT_INTERNAL_ERROR:
+		 /* Fall through */
+	default:
+		LOG_ERR("The modem encountered a fatal error during DFU: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Reinitializing nRF Modem Library");
+
+	ret = nrf_modem_lib_init(NORMAL_MODE);
+	if (ret) {
+		LOG_ERR("Reinitialization of nRF Modem library failed, retval: %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void on_pdn_activated(void)
+{
+	int ret;
+
+	ret = ipv4_addr_add(iface_bound);
+	if (ret == -ENODATA) {
+		LOG_WRN("No IPv4 address given by the network");
+		return;
+	} else if (ret) {
+		LOG_ERR("ipv4_addr_add, error: %d", ret);
+		net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+		nrf91_net_l2_disconnect(NULL);
+		return;
+	}
+
+	LOG_DBG("PDN IPv4 up");
+
+	net_if_dormant_off(iface_bound);
+
+	/* Cancel ongoing LTE timeout work, if ongoing. */
+	k_work_cancel_delayable(&lte_timeout_work);
+}
+
+static void on_pdn_deactivated(void)
+{
+	int ret;
+
+	ret = ipv4_addr_remove(iface_bound);
+	if (ret) {
+		LOG_ERR("ipv4_addr_remove, error: %d", ret);
+		net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+		net_if_dormant_on(iface_bound);
+		return;
+	}
+
+	ret = ipv6_addr_remove(iface_bound);
+	if (ret) {
+		LOG_ERR("ipv6_addr_remove, error: %d", ret);
+		net_mgmt_event_notify(NET_EVENT_IF_CONNECTIVITY_FATAL_ERROR, iface_bound);
+		net_if_dormant_on(iface_bound);
+		return;
+	}
+
+	net_if_dormant_on(iface_bound);
+
+	/* If persistence is disabled, LTE is deactivated upon a lost connection.
+	 * Re-establishment is reliant on the application calling net_if_connect()
+	 * or net_if_up() if NET_IF_NO_AUTO_CONNECT is disabled to
+	 * re-establish the LTE connection.
+	 */
+	if (!net_if_get_conn_persistence(iface_bound)) {
```

Comment:
The modem will take care of this. As long as the LTE stack is enabled on the modem it will try to maintain the connection. Even upon a loss of coverage.

Reply:
Yes, the modem will keep maintaining a connection indefinitely, but that is the problem

Imagine the following scenario:

Timeout is set to 10 seconds, and the connectivity layer connects successfully within 3 seconds. The timeout task is cancelled.

Now, later, connection is lost, and it takes more than 10 seconds for it to be regained

What is "supposed" to happen is after that 10 seconds, the connection attempt should be cancelled (despite persistence being enabled), but, if I am not mistaken, your current revision will keep trying indefinitely until it succeeds

I put supposed in quotes because that is merely how the spec for L2AC is currently written, we could change that if everyone involved thinks it would be better for the timeout to only apply on the first attempt. But IMO it makes more sense for it to apply to all subsequent re-connection attempts as well

Reply:
Ok, I understand. It makes sense to reschedule here. I also think we should add some documentation on the fact that the modem has "built in" reconnection logic. And that there should be a very good reason why said logic should be overshadowed by something in the application.

Maybe we should add an option to disable the timeout and rely solely on the modem to recover a lost connection.
the timeout could be set to `-1` for instance to disable it.

Reply:
> Maybe we should add an option to disable the timeout and rely solely on the modem to recover a lost connection.
the timeout could be set to -1 for instance to disable it.

Oh, good point -- In fact, the L2AC spec calls for a timeout of zero to imply the timeout is disabled. I'm glad you mentioned it, because I forgot to check for that when reviewing

Reply:
I used `-1` to indicate that the timer is disabled as its a typical value found throughout NCS. I guess `0` is used as well, but I prefer `-1` as its slighly more explicit.

Reply:
Whatever we do, it's important that we are consistent across connectivity implementations, so if -1 is going to be used, we need to bring this up on the API PR